src/governance.cpp

```diff
@@ -550,15 +525,39 @@ void CGovernanceManager::UpdateCachesAndClean()
                     ++lit;
                 }
             }
-            if(pObj->nObjectType == GOVERNANCE_OBJECT_WATCHDOG) {
-                mapWatchdogObjects.erase(it->first);
+
+            int nStatus;
+
+            int64_t nSuperblockCycleSeconds = Params().GetConsensus().nSuperblockCycle * Params().GetConsensus().nPowTargetSpacing;
+            int64_t nTimeExpired = pObj->GetCreationTime() + 2 * nSuperblockCycleSeconds + GOVERNANCE_DELETION_DELAY;
+
+            if(pObj->GetObjectType() == GOVERNANCE_OBJECT_WATCHDOG) {
+                mapWatchdogObjects.erase(nHash);
+                nStatus = (int)nTimeExpired;
+            } else if(pObj->GetObjectType() == GOVERNANCE_OBJECT_TRIGGER) {
+                nStatus = (int)nTimeExpired;
+            } else {
+                // keep hashes of deleted proposals forever
+                nStatus = std::numeric_limits<int>::max();
             }
+
+            mapSeenGovernanceObjects.insert(std::make_pair(nHash, nStatus));
             mapObjects.erase(it++);
         } else {
             ++it;
         }
     }
 
+    // forget about expired deleted objects
```

Comment:
yes, we can completely forget about triggers and watchdogs, created earlier than 2 months ago (2 * nSuperblockCycleSeconds). But proposals' hashes are saved forever:
https://github.com/the-schnibble/dash/blob/e51db9a6db63fe777f78d3b1270c21894e36644d/src/governance.cpp#L535